peachjam/models/user_following.py

```diff
@@ -0,0 +1,244 @@
+from countries_plus.models import Country
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.contrib.sites.models import Site
+from django.core.exceptions import ValidationError
+from django.core.mail import send_mail
+from django.db import models
+from django.template.loader import render_to_string
+from django.utils.translation import gettext_lazy as _
+
+from . import Author, CoreDocument, Court, CourtClass, CourtRegistry, Locality, Taxonomy
+
+
+class UserFollowing(models.Model):
+    user = models.ForeignKey(
+        get_user_model(),
+        on_delete=models.CASCADE,
+        related_name="following",
+        verbose_name=_("user"),
+    )
+    court = models.ForeignKey(
+        Court,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="followers",
+        verbose_name=_("court"),
+    )
+    author = models.ForeignKey(
+        Author,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="followers",
+        verbose_name=_("author"),
+    )
+    court_class = models.ForeignKey(
+        CourtClass,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="followers",
+        verbose_name=_("court class"),
+    )
+    court_registry = models.ForeignKey(
+        CourtRegistry,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="followers",
+        verbose_name=_("court registry"),
+    )
+    country = models.ForeignKey(
+        Country,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="followers",
+        verbose_name=_("country"),
+    )
+    locality = models.ForeignKey(
+        Locality,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="followers",
+        verbose_name=_("locality"),
+    )
+    taxonomy_topic = models.ForeignKey(
+        Taxonomy,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="followers",
+        verbose_name=_("taxonomy topic"),
+    )
+    last_alerted_at = models.DateTimeField(_("last alerted at"), null=True, blank=True)
+
+    created_at = models.DateTimeField(_("created at"), auto_now_add=True)
+
+    class Meta:
+        constraints = [
```

Comment:
Perhaps a unique constraint across all fields?

Reply:
each followed object has to be unique together with a user. unless i'm not getting what you mean?

Reply:
the constraint `(user, court, author, court_class, ...)` etc. should do that for you.

Reply:
hmm this wont work perfectly, since it will enforce uniqueness across all values combined.
eg (userA, courtA, authorA) and (userA, courtA, courtB) will be allowed.

Reply:
True, but we will never create objects like that. What we want to guard against is the user clicking "subscribe" very quickly (perhaps by mistake) and us creating two entries for the same court. Only one of the FK items will ever be non-null.

Reply:
id rather leave the constraints explicit like this. it better reflects how we want the model to behave